datumaro/plugins/synthetic_data/image_generator.py

```diff
@@ -0,0 +1,208 @@
+        )
+        parser.add_argument(
+            "--shape",
+            nargs="+",
+            type=int,
+            required=True,
+            help="Data shape. For example, for image with height = 256 and width = 224: --shape 256 224",
+        )
+
+        return parser
+
+    def __init__(self, output_dir: str, count: int, shape: Tuple[int, int]):
+        self._count = count
+        self._cpu_count = min(os.cpu_count(), self._count)
+        assert len(shape) == 2
+        self._height, self._width = shape
+
+        self._output_dir = output_dir
+        if not osp.exists(output_dir):
+            os.makedirs(output_dir)
+
+        if self._height < 13 or self._width < 13:
+            raise ValueError(
+                "Image generation with height or width of less than 13 is not supported"
+            )
+
+        self._weights = self._create_weights(IFSFunction.NUM_PARAMS)
+        self._threshold = 0.2
+        self._iterations = 200000
+
+        self._path = os.getcwd()
+        log.info(
+            "Downloading colorization model to '%s'",
+            self._path,
+        )
+        download_colorization_model(self._path)
+        self._initialize_params()
+
+    def generate_dataset(self) -> None:
+        log.info(
+            "Generating 3-channel images with height = '%d' and width = '%d'",
+            self._height,
+            self._width,
+        )
+
+        params = []
+        for i in range(self._categories):
+            params.append(self._generate_category(Random(i)))
+
+        # with Pool(processes=self._cpu_count) as pool:
+        #     params = pool.map(self._generate_category, [Random(i) for i in range(self._categories)])
+
+        instances_weights = np.repeat(self._weights, self._instances, axis=0)
+        weight_per_img = np.tile(instances_weights, (self._categories, 1))
+        repeated_params = np.repeat(params, self._weights.shape[0] * self._instances, axis=0)
+        repeated_params = repeated_params[: self._count]
+        weight_per_img = weight_per_img[: self._count]
+        assert weight_per_img.shape[0] == len(repeated_params) == self._count
+
+        splits = min(self._cpu_count, self._count)
+        params_per_proc = np.array_split(repeated_params, splits)
+        weights_per_proc = np.array_split(weight_per_img, splits)
+
+        generation_params = []
+        offset = 0
+        for i, (param, w) in enumerate(zip(params_per_proc, weights_per_proc)):
+            indices = list(range(offset, offset + len(param)))
+            offset += len(param)
+            generation_params.append((Random(i), param, w, indices))
+
+        # for i, param in enumerate(generation_params):
+        #     self._generate_image_batch(*param)
+
+        with Pool(processes=self._cpu_count) as pool:
+            pool.starmap(self._generate_image_batch, generation_params)
+        pool.close()
+        pool.join()
```

Comment:
I don't think these are necessary, as `Pool`'s `__exit__` implementation already calls `terminate()`.

Reply:
Yes, but hang on macos relates to multiprocessing.  Can I disable multiprocessing for macos and python < 3.8?

Reply:
I can suggest you:
- Try to launch a single-process version an test
- Make the processing function a staticmethod / classmethod
- Note, that `pool.*map*_async()` may not execute functions until we iterate over the result

Also, I've seen some messages on the CI page telling about reaching maximum count of something (cpu or processes) for a job. Maybe, it is related to the problem.